### ISSUE TITLE
Deprecate `R.arity` and/or `R.nAry`

### ISSUE THREAD

**Issue by paldepind** [ramda/ramda#1107]
I think at least one of these functions should be deprecated.

`R.nAry` is just `R.curryN` that doesn't curry. But when you have a fixed arity function there's no reason not to curry. It creates confusion. The first time (and only) time I used `R.nAry` it I expected it to curry and thus introduced a bug into my program.

`R.arity` seems really funky to me. It pretends to create a fixed arity function. But it's not really fixed since it can be abused by passing extra arguments. How is that going to help with anything but writing contrived code?

**Comment by davidchambers** [comment]
> `R.nAry` is just `R.curryN` that doesn't curry. But when you have a fixed arity function there's no reason not to curry.

I find this argument persuasive. Put so clearly, I see no reason to keep `R.nAry` (or the partially applied variants).

**Comment by buzzdecafe** [comment]
I also agree.

**Comment by CrossEye** [comment]
> `R.nAry` is just `R.curryN` that doesn't curry.

`nAry` is more than that.  It also _reduces_ the arguments passed to the underlying function:

    ['1', '2', '3'].map(parseInt); //=> [1, NaN, NaN]
    ['1', '2', '3'].map(R.nAry(1, parseInt)); //=> [1, 2, 3]

(And obviously, one would usually use `unary` for that.)

Whether or not it should also curry is another interesting question, but this does do something more than `curryN`.

`arity` is much less important.  I wrote it because I needed it internally.  I don't remember where now.  It seemed to do no harm to expose it.  But that was a stupid attitude.  Of course it should only have been exposed if there was a real need, not only if it did no harm.  If there is no internal need, I'm all for removing it altogether.  And even if there is, I don't think there would be much objection to removing it from the public interface.  The point of this one, though, was to report the arity properly to be passed to `curry`, but still to allow any arguments to pass through.  It is a bit of a Frankenfunction,

**Comment by paldepind** [comment]
> `nAry` is more than that. It also reduces the arguments passed to the underlying function:

Oh. I thought `R.curryN` did that as well.

But then it's `R.arity` that's just `R.curryN` without the currying.

**Comment by paldepind** [comment]
Shouldn't `curryN` limit the number of arguments passed to the function like `nAry` does? Or is there a use case for currying a function to a "fake" arity and pass it additional arguments beyond what the fake arity suggests? That would achieve some sort of currying on a variadic function. To me that sounds like an odd thing to do as it would break some of the rules that generally holds for curried functions. Like `fn(1)(2, 3) === fn(1)(2)(3)`.

If `curryN` limited the number of arguments passed to the curried function `nAry` could be deprecated and the upgrade path would simply be to replace `nAry` with `curryN`.

Incidentally implementing this is actually very simple after 50955c0.

**Comment by CrossEye** [comment]
> Shouldn't curryN limit the number of arguments passed to the function like nAry does? 

I think an argument would need to be made for that.  We are working with Javascript, where variadic functions are common, and there is a very common pattern of including a final `options` argument at the end of the parameter list.

So I don't think of this so much as a "fake arity" as a limiting arity.  The function won't be called until you supply _at least n_ arguments.  If at that time you also supply the _(n + 1)st_, the _(n + 2)nd_, etc. arguments, ISTM that one would want them passed along as well.

I'd suggest that instead of thinking of this as the "currying of a variadic function", it's instead allowing variadic application of a curried function.  The difference is perhaps subtle, but it's real.  

Here's a potential use: http://bit.ly/1cgCxnW


> [I]t would break some of the rules that generally holds for curried functions. Like `fn(1)(2, 3) === fn(1)(2)(3)`.

This is more concerning to me.  We made that declaration around #836, without considering this inconsistency.  We might be able to patch up the declaration to say something about a function curried to a particular length, but that loses some important simplicity.

I'm really not sure how to handle this.

**Comment by paldepind** [comment]
> I think an argument would need to be made for that. We are working with Javascript, where variadic functions are common, and there is a very common pattern of including a final options argument at the end of the parameter list.

Yes. That's correct. But we are also working with curried functions. And I don't think curried and variadic functions mix very well. If one wants to curry a variadic functions I think one should do one of two things:

1/ Don't curry it.
2/ Make the function invariadic.

I think currying is so powerful that is replaces the need for variadic functions in most use cases.

Consider your function:

```javascript
var sortOn = R.curryN(2, function(comparator, prop, descend) {
  return function(list) {
   	return R.sort(function(a, b) {
      return (descend ? -1 : 1) * comparator(a[prop], b[prop]);
    }, list) 
  }
});
```

What is the argument to `sortOn` that is least likely to change? That is `descend`. So you should put it first! But because you wanted an optional argument you had to put it last. Optional arguments pulls in the exact opposite direction of currying.

If you want currying then embrace it and do something this:

```javascript
var sortOn = R.curryN(3, function(descend, comparator, prop) {
  return function(list) {
   	return R.sort(function(a, b) {
      return (descend ? -1 : 1) * comparator(a[prop], b[prop]);
    }, list) 
  }
});
var sortAscOn = sortOn(false);
var sortDescOn = sortOn(true);
```

> So I don't think of this so much as a "fake arity" as a limiting arity. The function won't be called until you supply at least n arguments. If at that time you also supply the (n + 1)st, the (n + 2)nd, etc. arguments, ISTM that one would want them passed along as well.

> I'd suggest that instead of thinking of this as the "currying of a variadic function", it's instead allowing variadic application of a curried function. The difference is perhaps subtle, but it's real.

I can see what you mean. That is a way to think about currying. But, we're not talking about normal proper currying then.

> This is more concerning to me. We made that declaration around #836, without considering this inconsistency. We might be able to patch up the declaration to say something about a function curried to a particular length, but that loses some important simplicity.

I had actually forgotten that these rules are stated in the Ramda documentation. These rules goes beyond Ramda. They are (I think) part of the definition of what currying means.

Don't get me wrong. Variadic functions have uses IMO but mixing them with currying causes confusion and reduces the transparency of ones code.

Something else: Can we at least agree that `R.nAry` should curry the function it returns?

**Comment by buzzdecafe** [comment]
> Can we at least agree that `R.nAry` should curry the function it returns?

I agree, although that pretty much means it is `curryN`

**Comment by paldepind** [comment]
> I agree, although that pretty much means it is curryN

Excactly ;)

**Comment by buzzdecafe** [comment]
... and to support variadic functions we could try the approach recommended way back when in https://github.com/ramda/ramda/pull/138

**Comment by CrossEye** [comment]
>> Can we at least agree that R.nAry should curry the function it returns?

> I agree, although that pretty much means it is curryN

I don't have any objection to this currying its result.  But there are still some differences from `curryN`:  http://bit.ly/1Ay7xvq

**Comment by CrossEye** [comment]
> But we are also working with curried functions. And I don't think curried and variadic functions mix very well. If one wants to curry a variadic functions I think one should do one of two things:
> 
> 1/ Don't curry it.
> 2/ Make the function invariadic.

I'm starting to feel fairly lonely in the stand I'm taking, but I do still feel that Ramda should not be particularly prescriptive about how users write their Javsacript.  To my mind, it's supposed to be a library that works with the grain of the language.  It's trying to make easier a style of coding through functional composition, and to offer a number of standard (and less standard) functional tools, and by doing so, hopes to promote stronger functional practices.  But it's not designed to dictate too strongly how users could and should write their code.

That has a number of consequences that might be slowing us down in ways that others want to move Ramda.  There have been a lot of suggestions about currying the result of a number of Ramda functions.  `nAry` won't bother me, as I cannot see any possible harm.  But in many other cases, I hesitate because of the _named-optional-parameter_ issue.  If a user's `compose` returns such a function, and we auto-curry it, that function is hopelessly broken; the optional function becomes mandatory.  Similarly here, if a user has the function I supplied above -- simplified from something I was looking to use, although eventually I gave up for reaons like this -- I cannot wrap it up with `curryN`, if we alter the function.  I will have to write more intrusive wrappers just to use someone else's API the way I like.  And if I want to write APIs for use with Ramda, I have to think about all sorts of things Ramda's way, instead of the way I've thought about them for years in JS, such as the simple: _optional parameters come at the end._

So, while I agree that variadic funcitons and currying don't play well together, I think that means more that Ramda shouldn't be including variadic functions.  It doesn't mean that it should dictate that all user functions passed to it must be non-variadic.  Where we can comfortably support variadic functions, I see no objection.

----

>>> [I]t would break some of the rules that generally holds for curried functions. Like `fn(1)(2, 3) === fn(1)(2)(3)`.

>> This is more concerning to me. We made that declaration around #836, without considering this inconsistency.

> I had actually forgotten that these rules are stated in the Ramda documentation. These rules goes beyond Ramda. They are (I think) part of the definition of what currying means.

I've thought about this a bit more, and I'm much less concerned.  Such rules, only extend out to the arity of the function.  There is no expectation that this would be true:

    var fn = function(a, b) {return a + b;};
    fn(1, 2, 3) === fn(1, 2)(3); // the RHS is incoherent

**Comment by davidchambers** [comment]
What if `R.curryN` took *two* integer arguments: `min` and `max`? `min` would specify the number of arguments which must be provided before the wrapped function is applied; `max` would specify the maximum number of arguments forwarded to the wrapped function.

```javascript
['1', '2', '3'].map(R.curryN(1, 1, parseInt));
// => [1, 2, 3]

// Curried binary function which accepts any number
// of addition arguments as it is fully applied.
var compose = R.curryN(2, Infinity, R.compose);

compose(f, g);
compose(f)(g);
compose(f, g, h);
compose(f)(g, h);

// Curried binary function with upper bound.
var quux = R.curryN(2, 3, function(foo, bar, options) { ... });

quux(a, b);
quux(a)(b);
quux(a, b, c);
quux(a)(b, c);
quux(a, b, c, d, e);  // `d` and `e` will not be forwarded
quux(a)(b, c, d, e);  // `d` and `e` will not be forwarded
```

The implementation of `R.curry` would become:

```javascript
module.exports = _curry1(function curry(fn) {
  return curryN(fn.length, fn.length, fn);
});
```

**Comment by CrossEye** [comment]
> What if R.curryN took two integer arguments: `min` and `max`?

Although that might be an interesting API, it might be better as a new function.


```js
    var curryN = newFunc(__, Infinity); // current behavior
    var curryN = function(n, fn) {return newFunc(n, n, fn);}; // Simon's suggestion
```

**Comment by davidchambers** [comment]
> Although that might be an interesting API, it might be better as a new function.

What would we call it? `R.curryNN`?

My thinking is that `R.curryN` already exists to handle an edge case: functions which do not report their arity correctly. I use `R.curry` ten times for every time I require `R.curryN`. It seems like overkill to have `R.curry`, the lesser-used `R.curryN`, and the even lesser-used `R.curryNN`. I'd prefer to have one function for the common case; one function for the edge cases.

**Comment by CrossEye** [comment]
Yes, you're right.  There is no reason for it -- if we choose to go this way -- not to replace the existing one.  Because I've been so caught in discussions about `curryN`, I was forgetting how much of an edge case it already was.

**Comment by paldepind** [comment]
> I'm starting to feel fairly lonely in the stand I'm taking, but I do still feel that Ramda should not be particularly prescriptive about how users write their Javsacript.

I agree. But if a user _chooses_ to curry one of their functions they should take the consequence that having that function take an optional parameter doesn't make much sense. Users could still write all the variadic functions they want.

> I cannot wrap it up with curryN, if we alter the function. I will have to write more intrusive wrappers just to use someone else's API the way I like. And if I want to write APIs for use with Ramda, I have to think about all sorts of things Ramda's way, instead of the way I've thought about them for years in JS, such as the simple: optional parameters come at the end.

Isn't all of that already the case? Ramda already does a lot of things differently from normal JavaScript that doesn't fit in.

> What if R.curryN took two integer arguments: min and max? min would specify the number of arguments which must be provided before the wrapped function is applied; max would specify the maximum number of arguments forwarded to the wrapped function.

Why? That sound's overly complicated to me. Then I'd prefer to just keep the current behavior. Actually the current `curryN` behavior is not a _practical_ problem to me. I've never thought of mixing currying and optional arguments anyway. So I've never noticed the difference in behavior.

Can anyone show a use case for `R.nAry` beyond the often mentioned troublesome combination of `Array.prototype.map` and `parseInt`? That might reveal some cases where the current `curryN` could cause trouble.

**Comment by paldepind** [comment]
> The implementation of R.curry would become:
```javascript
module.exports = _curry1(function curry(fn) {
  return curryN(fn.length, fn.length, fn);
});
```

Then you're implementing the version of `curry` that I argue for. Not the current `curry` behavior.

**Comment by davidchambers** [comment]
> Then you're implementing the version of `curry` that I argue for. Not the current `curry` behavior.

Correct. Your version is preferable. :)

If a user wishes to curry a function which takes an optional argument, `R.curryN` would satisfy this requirement.

**Comment by CrossEye** [comment]
>> Then you're implementing the version of `curry` that I argue for. Not the current `curry` behavior.

> Correct. Your version is preferable. :)

I really have no problem with the main `curry` function doing what we all think would be the preferable behavior, so long as we have a fallback in something like `curryN` that would still work with variadic functions.

**Comment by CrossEye** [comment]
> But if a user _chooses_ to curry one of their functions they should take the consequence that having that function take an optional parameter doesn't make much sense. Users could still write all the variadic functions they want.

I meant to make the stronger stance, though, that Ramda should still be able to work with functions like this.  I mean that we should not be overly restrictive in what we accept.

This may sound contradictory, for I am also the one who pushes back anytime anyone suggests we do anything to work with sparse arrays, but my main feeling on them is that they are not really in the mainstream of JS development, just an oddity that the language supports.  Variadic functions, OTOH, are a core feature.

> Isn't all of that already the case? Ramda already does a lot of things differently from normal JavaScript that doesn't fit in.

And I do hope that Ramda helps a certain type of developer learn to develop in a different style.  But I would rather do that gradually, with the user able to slowly increase her FP usage, rather than as a one-shot tool that offers a wonderful new paradigm, but only if you're willing to drink a lot of our Kool-Aid.

> Can anyone show a use case for `R.nAry` beyond the often mentioned troublesome combination of `Array.prototype.map` and `parseInt`? That might reveal some cases where the current `curryN` could cause trouble.

I don't have an existing one handy.  But I can create one:

```js
var makeRect = (function() {
  var defaults = {
    color: 'transparent'
  };
  return function makeRect(w, h, opts) {
    var width = w;
    var height = arguments.length > 1 ? h : w;
    var props = R.merge(defaults, opts || {});
    return {
      width: width,
      height: height,
      color: props.color,
    };
  };
}());

var simpleRect = R.nAry(2, makeRect);

makeRect(5, 3), //=> {"color": "transparent", "height": 3,, "width": 5} 
makeRect(4, 7, {color: 'red'}), //=> {"color": "red", "height": 7, "width": 4}]
makeRect(6), //=> {"color": "transparent", "height": 6, "width": 6}
simpleRect(10, 20), //=> {"color": "transparent", "height": 20, "width": 10}
simpleRect(15), //=> NOTE: an incomplete rectangle, height not supplied
  // NOTE: options are ignored
simpleRect(30, 40, {color: 'blue'}), //=> {"color": "transparent", "height": 40, "width": 30} 
```

Note that `nAry` simplifies this fairly complex function of one, two, or three parameters, and fixes it at  two.  Currying the result might not hurt, but it doesn't do anything useful either.

I'm not trying to claim anything nice about this function, just that I would like to be able to support code that's written this way, if at all possible.

**Comment by paldepind** [comment]
> I meant to make the stronger stance, though, that Ramda should still be able to work with functions like this. I mean that we should not be overly restrictive in what we accept.

Ok. I get it. The hard part for me to understand is that some people would want to curry variadic functions. That has just never been the way I though about currying.

But after thinking about it I can understand your way of seeing `curryN` as currying to a "limiting arity". And what happens beyond that limiting arity is up to the curried function. I still don't think it makes much sense when one thinks about what currying _actually is_. But it does make sense considering the sloppy way that arguments are generally handled in JavaScript and that it has practical benefits if one is willing to relax ones understanding of currying.

What I'm trying to say is that I now agree with you. But only if passing along the additional arguments can never do harm to people using curry in the more strict sense. And I cannot currently think of any such harm.

I do however find your `R.nArity` example very contrived. You first deliberately create a version of a variadic function that doesn't accept an optional parameter. Why would I ever do that? And then you go pass in that optional parameter anyway. That is surely a mistake and the programmer is not getting the behavior he expects in either case?

To sum it up I'm currently fine with `curry` and `curryN` being slightly the way they are. But I still don't see much use of `nArity`.

**Comment by CrossEye** [comment]
> I do however find your R.nArity example very contrived. You first deliberately create a version of a variadic function that doesn't accept an optional parameter. Why would I ever do that? 

The idea was to show that I can create a different API out of the existing one using `nAry`.  The new API accepts two and only two parameters.  The idea of showing the other two calls was to show that they wouldn't work with the new API: one _should never_ call this new function that way.
